File: packages/python/src/mpesa/services/b2b.py
from typing import Optional
from .http.http_client import HttpClient
from .auth import Auth
from .utils import Helpers


class B2BService:
    def __init__(self, http_client: HttpClient, auth: Auth, config):
        self.http_client = http_client
        self.auth = auth
        self.helpers = Helpers(config)
        self.base_url = http_client.get_base_url()

    def send(self, receiver_shortcode: str, command_id: str, amount: int, remarks: str,
             account_number: Optional[str] = None, result_url: Optional[str] = None,
             timeout_url: Optional[str] = None, short_code_type: str = 'B2B') -> dict:
        if command_id == 'BusinessPayBill' and not account_number:
            raise ValueError('Account Number is required for BusinessPayBill CommandID')
        
        url = f'{self.base_url}/mpesa/b2b/v1/paymentrequest'
        
        body = {
            'Initiator': self.helpers.get_config('initiator_name'),
            'SecurityCredential': self.helpers.generate_security_credential(),
            'CommandID': command_id,
            'SenderIdentifierType': '4',
            'RecieverIdentifierType': '4',
            'Amount': amount,
            'PartyA': self.helpers.get_config('b2c_shortcode'),
            'PartyB': receiver_shortcode,
            'AccountReference': account_number,
            'Remarks': remarks,
            'ResultURL': result_url or self.helpers.get_config('callbacks.b2b_result_url'),
            'QueueTimeOutURL': timeout_url or self.helpers.get_config('callbacks.b2b_timeout_url'),
        }
        
        token = self.auth.get_access_token(short_code_type)
        return self.http_client.post(url, body, token)

File: packages/node/src/services/b2b.ts
import { B2bParams } from '../types';
import { HttpClient } from '../http/httpClient';
import { Auth } from '../core/Auth';
import { Helpers } from '../utils/helpers';

export class B2BService {
  private httpClient: HttpClient;
  private auth: Auth;
  private helpers: Helpers;
  private baseUrl: string;

  constructor(httpClient: HttpClient, auth: Auth) {
    this.httpClient = httpClient;
    this.auth = auth;
    this.helpers = new Helpers();
    this.baseUrl = httpClient.getBaseUrl();
  }

  async send(params: B2bParams): Promise<any> {
    const { receiverShortcode, commandId, amount, remarks, accountNumber, resultUrl, timeoutUrl, shortCodeType = 'B2B' } = params;

    if (commandId === 'BusinessPayBill' && !accountNumber) {
      throw new Error('Account Number is required for BusinessPayBill CommandID');
    }

    const url = `${this.baseUrl}/mpesa/b2b/v1/paymentrequest`;
    const body = {
      Initiator: this.helpers.getConfig('initiatorName'),
      SecurityCredential: this.helpers.generateSecurityCredential(),
      CommandID: commandId,
      SenderIdentifierType: '4',
      RecieverIdentifierType: '4',
      Amount: amount,
      PartyA: this.helpers.getConfig('b2cShortcode'),
      PartyB: receiverShortcode,
      AccountReference: accountNumber,
      Remarks: remarks,
      ResultURL: resultUrl || this.helpers.getConfig('callbacks.b2bResultUrl'),
      QueueTimeOutURL: timeoutUrl || this.helpers.getConfig('callbacks.b2bTimeoutUrl'),
    };

    const token = await this.auth.getAccessToken(shortCodeType);
    return this.httpClient.post(url, body, token);
  }
}
